### ISSUE TITLE
XMS-CASPT2 Geometry Optimization

### ISSUE THREAD

**Issue by meaganoakley** [nubakery/bagel#183]
Hi,

I am having some issues with trying to optimize geometry with XMS-CASPT2. It will complete the CASSCF iterations, but will fail when trying to start optimization. Here is the test case I have been working on: 

### Input:
{ "bagel" : [

{
  "title" : "molecule",
  "basis" : "svp",
  "df_basis" : "svp-jkfit",
  "geometry" : [
    { "atom" : "Li", "xyz" : [ 0.000000, 0.000000, 6.000000] },
    { "atom" : "F",  "xyz" : [ 0.000000, 0.000000, 0.000000] }
  ]
},

{
  "title" : "optimize",
  "explicitbond" : true,
  "explicit" : [
 { "pair" : [1, 2]}
 ],
 "target" : 0,
  "numerical" : true,
  "method" : [ {
    "title" : "casscf",
    "nstate" : 4,
    "nact" : 4,
    "nclosed" : 3
  },

  { "title" : "smith",
    "method" : "caspt2",
    "ms" : "true",
    "xms" : "true",
    "sssr" : "true",
    "shift" : 0.2,
    "frozen" : true
  } ]
}

]}

### Output: 
 === SMITH program ===

    * freezing 2 orbitals
    * SS-SR internal contraction is used

    * Approximate memory requirement for SMITH calculation per MPI process:
      o Storage requirement for T-amplitude, lambda, and residual is 0.00 GB
      o Storage requirement for MO integrals is 0.00 GB

  *** Geometry ***

  { "atom" : "Li", "xyz" : [  ERROR ON MPI PROCESS 20: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)      0.000000,      0.000000,  ERROR ON MPI PROCESS 12: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)  ERROR ON MPI PROCESS       6.000000  ERROR ON MPI PROCESS 13: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 22: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 14: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 23: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 15: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)

  ERROR ON MPI PROCESS 21  ERROR ON MPI PROCESS 17: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
: EXCEPTION RAISED:
a key was not found in Storage::locate(const size_t key)16
: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
 ] },
  { "atom" : "F", "xyz" : [      0.000000  ERROR ON MPI PROCESS 4: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key),  ERROR ON MPI PROCESS 5: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 6: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 7: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 8: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 9: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS   ERROR ON MPI PROCESS 11: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)

10: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
      0.000000  ERROR ON MPI PROCESS 2: EXCEPTION RAISED:    ERROR ON MPI PROCESS 3: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)a key was not found in Storage::locate(const size_t key)

,      0.000000 ]  ERROR ON MPI PROCESS 1: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
 },

  ERROR ON MPI PROCESS 0: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)
  ERROR ON MPI PROCESS 18: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)  ERROR ON MPI PROCESS 19: EXCEPTION RAISED:  a key was not found in Storage::locate(const size_t key)

I have been working with staff at the HPC institute I work with and they came back with this: 
The fatal error occurs on line 53 of bagel/src/grad/finite.cc, where there is a divide by zero error on the final pass of the 'for' loop. The error occurs when (natom * 3) is evenly divisible by ncomm, which will occur depending on the number of atoms in your system, the number of MPI threads you request, and the number of processors available to the software. It looks like there needs to be a check of the value of ncolor before an attempt to divide by it, and choose a different behaviour if it is equal to zero.

Thank you in advance for the help,

Meagan

**Comment by shiozaki** [comment]
Hi Meagan

For technical reasons, the XMS-CASPT2 optimization has to be done with a slightly different input. Please see below (I didn't check so you may need to adjust a bit). Please use analytical gradients. There is also an example in https://nubakery.org/opt/optimize.html
```
{
  "title" : "casscf",
  "nstate" : 4,
  "nact" : 4,
  "nclosed" : 3
},
{
  "title" : "optimize",
  "explicitbond" : true,
  "explicit" : [
    { "pair" : [1, 2]}
  ],
  "target" : 0,
  "numerical" : false,
  "method" : [ {
      "title" : "caspt2",
      "smith" : {
        "ms" : "true",
        "xms" : "true",
        "sssr" : "true",
        "shift" : 0.2,
        "frozen" : true
      }
    } ]
}
```

**Comment by shiozaki** [comment]
@jwpk1201 
This is a bug that is related to the one I mentioned. Please see below:

```
I have been working with staff at the HPC institute I work with and they came back with this:
The fatal error occurs on line 53 of bagel/src/grad/finite.cc, where there is a divide by zero error on the final pass of the 'for' loop. The error occurs when (natom * 3) is evenly divisible by ncomm, which will occur depending on the number of atoms in your system, the number of MPI threads you request, and the number of processors available to the software. It looks like there needs to be a check of the value of ncolor before an attempt to divide by it, and choose a different behaviour if it is equal to zero.
```

**Comment by shiozaki** [comment]
The ncolor issue should be fixed by 05ae88b108a372b5ef7e7290dc0e03f528c05367
